server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/filemove/FileMoveDetectionStep.java

```diff
@@ -195,32 +192,59 @@ public void visitFile(Component file) {
   }
 
   private ScoreMatrix computeScoreMatrix(Map<String, DbComponent> dtosByKey, Set<String> dbFileKeys, Map<String, File> reportFileSourcesByKey) {
-    int[][] scoreMatrix = new int[dbFileKeys.size()][reportFileSourcesByKey.size()];
+    ScoreMatrix.ScoreFile[] newFiles = reportFileSourcesByKey.entrySet().stream()
+      .map(e -> new ScoreMatrix.ScoreFile(e.getKey(), e.getValue().getLineCount()))
+      .toArray(ScoreMatrix.ScoreFile[]::new);
+    ScoreMatrix.ScoreFile[] removedFiles = dbFileKeys.stream()
+      .map(key -> {
+        DbComponent dbComponent = dtosByKey.get(key);
+        return new ScoreMatrix.ScoreFile(dbComponent.getKey(), dbComponent.getLineCount());
+      })
+      .toArray(ScoreMatrix.ScoreFile[]::new);
+    // sort by highest line count first
+    Comparator<ScoreMatrix.ScoreFile> c = (o1, o2) -> -1 * Integer.compare(o1.getLineCount(), o2.getLineCount());
+    Arrays.sort(newFiles, c);
+    Arrays.sort(removedFiles, c);
+    int[][] scoreMatrix = new int[removedFiles.length][newFiles.length];
+    int lastNewFileIndex = newFiles.length - 1;
     int maxScore = 0;
 
     try (DbSession dbSession = dbClient.openSession(false)) {
-      int dbFileIndex = 0;
-      for (String removedFileKey : dbFileKeys) {
-        File fileInDb = getFile(dbSession, dtosByKey.get(removedFileKey));
-        if (fileInDb == null) {
+      for (int removeFileIndex = 0; removeFileIndex < removedFiles.length; removeFileIndex++) {
+        ScoreMatrix.ScoreFile removedFile = removedFiles[removeFileIndex];
+        int lowerBound = (int) Math.floor(removedFile.getLineCount() * 0.84);
+        int upperBound = (int) Math.ceil(removedFile.getLineCount() * 1.18);
+        // short circuit if all files are out of bound
+        if (newFiles[0].getLineCount() <= lowerBound || newFiles[lastNewFileIndex].getLineCount() >= upperBound) {
           continue;
         }
 
-        int reportFileIndex = 0;
-        for (Map.Entry<String, File> reportFileSourceAndKey : reportFileSourcesByKey.entrySet()) {
-          File unmatchedFile = reportFileSourceAndKey.getValue();
+        File fileInDb = null;
+        for (int newFileIndex = 0; newFileIndex <= lastNewFileIndex; newFileIndex++) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Reduce the total number of break and continue statements in this loop to use at most one. ([squid:S135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10253%2Foptimize_file_move_detection_based_on_file_size&issues=AWGU6531HYoMxGt2NSWQ&open=AWGU6531HYoMxGt2NSWQ)

[ISSUE_KEY]: AWGU6531HYoMxGt2NSWQ